pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+an expression of the form ``BinOp(UnaryOp('-', 3), '+', 4j)``).
+Interpolated *f*-strings, on the other hand, are not literal values, despite
+their appearance and can therefore not be used as literal patterns.
+
+Literal patterns not only occur as patterns in their own right, but also
+as keys in *mapping patterns*.
+
+
+.. _constant_value_pattern:
+
+Constant Value Patterns
+~~~~~~~~~~~~~~~~~~~~~~~
+
+It is good programming style to use named constants for parametric values or
+to clarify the meaning of particular values.  Clearly, it would be desirable
+to also write ``case (HttpStatus.OK, body):`` rather than
+``case (200, body):``, say.  The main issue that arises here is how to
+discern capture patterns (variables) and constant value patterns.  The
+general discussion surrounding this issue has brought forward a plethora of
+options, which we cannot all fully list here.
+
+Strictly speaking, constant value patterns are not really necessary, but
+could be implemented using guards, i.e.
+``case (status, body) if status == HttpStatus.OK:``.  Nonetheless, the
+convenience of constant value patterns is unquestioned and obvious.
+
+The observation that constants tend to be written in uppercase letters or
+collected in enumeration-like namespaces suggests possible rules to discern
+constants syntactically.  However, the idea of using upper vs. lower case as
+a marker has been met with scepticism since there is no similar precedence
+in core Python (although it is common in other languages).  We therefore only
+adopted the rule that any dotted name (i.e. attribute access) is to be
+interpreted as a constant value pattern as exemplified by ``HttpStatus.OK``
+above.  This excludes, in particular, local variables from acting as
+constants.
+
+Global variables can only be directly used as constent when defined in other
+modules, although there are work arounds to access the current module as a
+namespace as well.  A proposed rule to use a leading dot (e.g.
+``.CONSTANT``) for that purpose was critisised because it was felt that the
+dot would not be a visible-enough marker for that purpose.  Partly inspired
+by use cases in other programming languages, a number of different
+markers/sigils was proposed (such as ``^CONSTANT``, ``$CONSTANT``,
+``==CONSTANT``, etc.), although there was no obvious or natural choice.
+The current proposal therefore leaves the discussion and possible
+introduction of such a 'constant' marker for future PEPs.
+
+Distinguishing the semantics of names based on whether it is a global
+variable (i.e. the compiler would treat global variables as constants rather
+than capture patterns) leads to various issues.  The addition or alteration
+of a global variable in the module could have unintended side effects on
+patterns.  Moreover, pattern matching could not be used directly inside a
+module's scope because all variables would be global, making capture
+patterns impossible.
```

Comment:
(There was an even worse idea, look whether the name lookup currently succeeds. Not sure it's even worth mentioning it given its flaws?)

Reply:
I am not quite sure I understand your comment here, sorry...

Reply:
The idea was to to a lookup of the name, and if it's a NameError, bind it, otherwise compare it.
```
x = 42
match f():
    case x: print("it's 42")
del x
match f():
    case x: print("captured", x)
```
Forget you even saw this. :-)

Reply:
Ah, OK, thanks.  Yes, I ignored that :-).  It feels a bit like a "random code" generator for all those who enjoy the excitement of unpredictability in their programs ;-).